README
BlueZ - Bluetooth protocol stack for Linux
******************************************

Copyright (C) 2000-2002  Maxim Krasnyansky <maxk@qualcomm.com>
Copyright (C) 2003-2006  Marcel Holtmann <marcel@holtmann.org>

Bluetooth packet analyzer


Compilation and installation
============================

In order to compile Bluetooth analyzer you need following software packages:
	- Linux Bluetooth protocol stack (BlueZ)
	- GCC compiler

To configure run:
	./configure --prefix=/usr --mandir=/usr/share/man
 
Configure automatically searches for all required components and packages. 

To compile and install run:
	make && make install


Information
===========

Mailing lists:
	bluez-users@lists.sf.net - BlueZ general questions and discussions
	bluez-devel@lists.sf.net - BlueZ development 

For additional information about the project visit BlueZ web site:
	http://www.bluez.org


ChangeLog
ver 1.40:
	Add HCI and LMP decoding support for Simple Pairing.

	Note:
	This version needs at least bluez-libs-3.14

ver 1.39:
	Add support for daemon discover protocol.
	Fix device disconnect handling.

ver 1.38:
	Add support for daemon mode.

ver 1.37:
	Add decoding for SCO packet flags.
	Add decoding for more extended inquiry features.
	Add missing HCI command and event decodings.

	Note:
	This version needs at least bluez-libs-3.12

ver 1.36:
	Add missing HCI command definitions.
	Add missing HCI event definitions.
	Add missing HCI error definitions.

ver 1.35:
	Add partial decoding for eSCO setup.
	Fix HCI command array size.

ver 1.34:
	Add reading support for MacOS X packet logger format.
	Add decoding for AFH host channel classification command.
	Add decoding for QoS setup command.

ver 1.33:
	Add support for IPv6 decoding.
	Add IPv6 connection support.

ver 1.32:
	Add decoding for link supervision timeout change event.
	Add decoding for host controller to host flow control setting.
	Add decoding for host number of completed packets.
	Add decoding for host buffer size command.
	Fix OBEX header parsing.

	Note:
	This version needs at least bluez-libs-3.3

ver 1.31:
	Add Ericsson LMP decoding support.
	Update sniff subrating decoding.
	Update UUID and attribute definitions.

	Note:
	This version needs at least bluez-libs-3.0

ver 1.30:
	Add decoding for pause encryption feature.
	Add generic RFCOMM streaming helpers.
	Add HDLC decoding support to the PPP parser.
	Add unslip support and basic protocol decoding.
	Add PPP extraction support.
	Move TCP/IP decoders into a separate file.
	Fix DoS problem of the L2CAP parser.

ver 1.29:
	Obfuscate the BD_ADDR if no vendor events are requested.
	Decode and display binary strings correctly.
	Allow null-terminated text strings.
	Add definitions for Apple Agent.
	Add skeleton for AVCTP parser.
	Add skeleton for PPP parser.

ver 1.28:
	Add missing UUID definitions and translations.
	Add option to disable vendor commands and events.
	Add decoding for hold mode, sniff mode and park state.
	Add decoding for sniff subrate command complete event.
	Decode class of device and BD_ADDR for inquiry filter.
	Make it possible to obfuscate pin codes and link keys.

	Note:
	This version needs at least bluez-libs-2.23

ver 1.27:
	Decode the extended inquiry response payload.
	Update HCI_DEV_NONE frame format.
	Update the CSR PS key handling.
	Add another bunch LMP message decodings.
	Add compile time buffer checks (FORTIFY SOURCE).

ver 1.26:
	Fix memory leak with lmp_vertostr() function.
	Fix FHS packet decoding for slave connections.
	Add support for decoding more LMP messages.
	Display the data of extended inquiry results.
	Correct HCI command and event string arrays.
	Add SCO audio extraction support.

ver 1.25:
	Add new attribute identifiers and UUID definitions.
	Add decoding support for extended inquiry response.
	Add more detailed BCCMD decoding.
	
	Note:
	This version needs at least bluez-libs-2.20

ver 1.24:
	Allow selection of system device.
	Handle system events and commands.

ver 1.23:
	Add decoding support for inquiry scan type.
	Fix connection accept timeout and scan enable decoding.
	Show human readable timestamps only in verbose decoding mode.
	Dump raw BNEP payload only in non-verbose decoding mode.
	Update CSR BCCMD support and correct uint32 decoding.

	Note:
	This version needs at least bluez-libs-2.18

ver 1.22:
	Only do verbose decoding when requested.
	Fix number of completed packets decoding.
	Improve the decoding of CSR vendor commands and events.
	Use human readable timestamps.
	Add support for the BTSnoop file format.

ver 1.21:
	Clear L2CAP states when receiving the HCI disconnect.
	Make the OBEX parser aware of that RFCOMM is a stream.
	Update HCI verbose decoding routines.
	Add CSR verbose decoding support.

	Note:
	This version needs at least bluez-libs-2.17

ver 1.20:
	Workaround for inquiry results with RSSI and page scan mode.
	Decode almost every used HCI commands and events.

	Note:
	This version needs at least bluez-libs-2.16

ver 1.19:
	Fix error message decoding.
	Add IAC decoding support.
	Add LMP parser support.
	Add BCCMD decoding support.
	Add L2CAP retransmission and flow control decoding.

ver 1.18:
	Fix declaration of the SDP data structures.
	Extend the HCI verbose decoding support.

ver 1.17:
	Add support for HCI verbose decoding.
	Add support for L2CAP verbose decoding.
	Add first version of the BPA sniffer utility.

	Note:
	This version needs at least bluez-libs-2.15

ver 1.16:
	Add UUID translation for WAP.
	Fix display of UUID 128.
	Fix display of L2CAP config hint bit and QoS option.
	Fix parsing of SDP continuation state.

ver 1.15:
	Update the L2CAP channel counting routine.
	Fix SBC codec specific decoding.
	Fix AVDTP signal channel header decoding.
	Add decoding for the AVDTP media channel header.

ver 1.14:
	Count the number of L2CAP channels per PSM.
	Differ between the AVDTP signal and media channel.
	Add full decoding of the AVDTP signal channel information.

ver 1.13:
	Update many UUID to text translations.
	Add support for sending and receiving dumps over TCP.
	Add support for the vendor packets of the Digianswer BPA.

	Note:
	This version needs at least bluez-libs-2.11

ver 1.12:
	Fix whitespace stuff for extended dump.
	Add support for dynamic RFCOMM channels.
	Add initial OBEX parser support.

ver 1.11:
	Trace company id from the chip manufacturer.
	Support extended dump functionality.
	Use the L2CAP_CONF_RFC_MODE value.
	Use bt_get_unaligned() for unaligned access.
	Make compiling with debug information optional.
	Don't override CFLAGS from configure.

	Note:
	This version needs at least bluez-libs-2.10

ver 1.10:
	Fix display of L2CAP CID.
	Show L2CAP config values only when they are present.
	Decode L2CAP information request and response.
	Update autoconf and automake routines.

ver 1.9:
	Initial AVDTP parser support.
	Various minor fixes and cleanups.

ver 1.8:

src/csrsniff.c
/*
 *
 *  BlueZ - Bluetooth protocol stack for Linux
 *
 *  Copyright (C) 2004-2006  Marcel Holtmann <marcel@holtmann.org>
 *
 *
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program; if not, write to the Free Software
 *  Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
 *
 */

#ifdef HAVE_CONFIG_H
#include <config.h>
#endif

#include <stdio.h>
#include <errno.h>
#include <unistd.h>
#include <stdlib.h>
#include <getopt.h>
#include <signal.h>
#include <sys/ioctl.h>
#include <sys/socket.h>

#include <bluetooth/bluetooth.h>
#include <bluetooth/hci.h>
#include <bluetooth/hci_lib.h>

static volatile sig_atomic_t __io_canceled = 0;

static void sig_hup(int sig)
{
}

static void sig_term(int sig)
{
	__io_canceled = 1;
}

static struct {
	uint16_t id;
	uint16_t ver;
	char *date;
} firmware_map[] = {
	{  195,  1, "2001-11-27"	},
	{  220,  2, "2002-01-03"	},
	{  269,  3, "2002-02-22"	},
	{  270,  4, "2002-02-26"	},
	{  284,  5, "2002-03-12"	},
	{  292,  6, "2002-03-20"	},
	{  305,  7, "2002-04-12"	},
	{  306,  8, "2002-04-12"	},
	{  343,  9, "2002-05-02"	},
	{  346, 10, "2002-05-03"	},
	{  355, 11, "2002-05-16"	},
	{  256, 11, "2002-05-16"	},
	{  390, 12, "2002-06-26"	},
	{  450, 13, "2002-08-16"	},
	{  451, 13, "2002-08-16"	},
	{  533, 14, "2002-10-11"	},
	{  580, 15, "2002-11-14"	},
	{  623, 16, "2002-12-12"	},
	{  678, 17, "2003-01-29"	},
	{  847, 18, "2003-04-17"	},
	{  876, 19, "2003-06-10"	},
	{  997, 22, "2003-09-05"	},
	{ 1027, 23, "2003-10-03"	},
	{ 1029, 24, "2003-10-03"	},
	{ 1112, 25, "2003-12-03"	},
	{ 1113, 25, "2003-12-03"	},
	{ 1133, 26, "2003-12-18"	},
	{ 1134, 26, "2003-12-18"	},
	{ 1223, 27, "2004-03-08"	},
	{ 1224, 27, "2004-03-08"	},
	{ 1319, 31, "2004-04-22"	},
	{ 1320, 31, "2004-04-22"	},
	{ 1427, 34, "2004-06-16"	},
	{ 1508, 35, "2004-07-19"	},
	{ 1509, 35, "2004-07-19"	},
	{ 1587, 36, "2004-08-18"	},
	{ 1588, 36, "2004-08-18"	},
	{ 1641, 37, "2004-09-16"	},
	{ 1642, 37, "2004-09-16"	},
	{ 1699, 38, "2004-10-07"	},
	{ 1700, 38, "2004-10-07"	},
	{ 1752, 39, "2004-11-02"	},
	{ 1753, 39, "2004-11-02"	},
	{ 1759, 40, "2004-11-03"	},
	{ 1760, 40, "2004-11-03"	},
	{ 1761, 40, "2004-11-03"	},
	{ 2009, 41, "2005-04-06"	},
	{ 2010, 41, "2005-04-06"	},
	{ 2011, 41, "2005-04-06"	},
	{ 2016, 42, "2005-04-11"	},
	{ 2017, 42, "2005-04-11"	},
	{ 2018, 42, "2005-04-11"	},
	{ 2023, 43, "2005-04-14"	},
	{ 2024, 43, "2005-04-14"	},
	{ 2025, 43, "2005-04-14"	},
	{ 2032, 44, "2005-04-18"	},
	{ 2033, 44, "2005-04-18"	},
	{ 2034, 44, "2005-04-18"	},
	{ 2288, 45, "2005-07-08"	},
	{ 2289, 45, "2005-07-08"	},
	{ 2290, 45, "2005-07-08"	},
	{ 2388, 46, "2005-08-17"	},
	{ 2389, 46, "2005-08-17"	},
	{ 2390, 46, "2005-08-17"	},
	{ 2869, 47, "2006-02-15"	},
	{ 2870, 47, "2006-02-15"	},
	{ 2871, 47, "2006-02-15"	},
	{ 3214, 48, "2006-02-16"	},
	{ 3215, 48, "2006-02-16"	},
	{ 3216, 48, "2006-02-16"	},
	{    0, }
};

static int id2ver(uint16_t id)
{
	int i;

	for (i = 0; firmware_map[i].id; i++)
		if (firmware_map[i].id == id)
			return firmware_map[i].ver;

	return -1;
}

static void usage(void)
{
	printf("csrsniff - Utility for the CSR BlueCore sniffers\n\n");
	printf("Usage:\n"
		"\tcsrsniff [-i <dev>] <master-bdaddr> [slave-bdaddr]\n");
}

static struct option main_options[] = {
	{ "help",	0, 0, 'h' },
	{ "device",	1, 0, 'i' },
	{ 0, 0, 0, 0}
};

int main(int argc, char *argv[])
{
	struct sigaction sa;
	struct hci_dev_info di;
	struct hci_version ver;
	struct hci_filter flt;
	bdaddr_t bdaddr, master, slave;
	int need_raw;
	int dd, opt, dev = 0;

	bacpy(&slave, BDADDR_ANY);

	while ((opt=getopt_long(argc, argv, "+i:h", main_options, NULL)) != -1) {
		switch (opt) {
		case 'i':
			dev = hci_devid(optarg);
			if (dev < 0) {
				perror("Invalid device");
				exit(1);
			}
			break;

		case 'h':
		default:
			usage();
			exit(0);
		}
	}

	argc -= optind;
	argv += optind;
	optind = 0;

	if (argc < 1) {
		usage();
		exit(1);
	}

	str2ba(argv[0], &master);

	if (argc > 1)
		str2ba(argv[1], &slave);

	dd = hci_open_dev(dev);
	if (dd < 0) {
		fprintf(stderr, "Can't open device hci%d: %s (%d)\n",
						dev, strerror(errno), errno);
		exit(1);
	}

	if (hci_devinfo(dev, &di) < 0) {
		fprintf(stderr, "Can't get device info for hci%d: %s (%d)\n",
						dev, strerror(errno), errno);
		hci_close_dev(dd);
		exit(1);
	}

	if (hci_read_local_version(dd, &ver, 1000) < 0) {
		fprintf(stderr, "Can't read version for hci%d: %s (%d)\n",
						dev, strerror(errno), errno);
		hci_close_dev(dd);
		exit(1);
	}

	if (ver.manufacturer != 10 || id2ver(ver.hci_rev) < 0) {
		fprintf(stderr, "Can't find sniffer at hci%d: %s (%d)\n",
						dev, strerror(ENOSYS), ENOSYS);
		hci_close_dev(dd);
		exit(1);
	}

	if (!bacmp(&di.bdaddr, BDADDR_ANY)) {
		if (hci_read_bd_addr(dd, &bdaddr, 1000) < 0) {
			fprintf(stderr, "Can't read address for hci%d: %s (%d)\n",
						dev, strerror(errno), errno);
			hci_close_dev(dd);
			exit(1);
		}
	} else
		bacpy(&bdaddr, &di.bdaddr);

	need_raw = !hci_test_bit(HCI_RAW, &di.flags);

	hci_filter_clear(&flt);
	hci_filter_set_ptype(HCI_ACLDATA_PKT, &flt);
	hci_filter_set_ptype(HCI_EVENT_PKT, &flt);
	hci_filter_set_event(EVT_VENDOR, &flt);

	if (setsockopt(dd, SOL_HCI, HCI_FILTER, &flt, sizeof(flt)) < 0) {
		fprintf(stderr, "Can't set filter for hci%d: %s (%d)\n",
						dev, strerror(errno), errno);
		hci_close_dev(dd);
		exit(1);
	}

	memset(&sa, 0, sizeof(sa));
	sa.sa_flags   = SA_NOCLDSTOP;
	sa.sa_handler = SIG_IGN;
	sigaction(SIGCHLD, &sa, NULL);
	sigaction(SIGPIPE, &sa, NULL);

	sa.sa_handler = sig_term;
	sigaction(SIGTERM, &sa, NULL);
	sigaction(SIGINT,  &sa, NULL);

	sa.sa_handler = sig_hup;
	sigaction(SIGHUP, &sa, NULL);

	if (need_raw) {
		if (ioctl(dd, HCISETRAW, 1) < 0) {
			fprintf(stderr, "Can't set raw mode on hci%d: %s (%d)\n",
						dev, strerror(errno), errno);
			hci_close_dev(dd);
			exit(1);
		}
	}

	printf("CSR sniffer - Bluetooth packet analyzer ver %s\n", VERSION);

	if (need_raw) {
		if (ioctl(dd, HCISETRAW, 0) < 0)
			fprintf(stderr, "Can't clear raw mode on hci%d: %s (%d)\n",
						dev, strerror(errno), errno);
	}

	hci_close_dev(dd);

	return 0;
}


src/bpasniff.c
/*
 *
 *  BlueZ - Bluetooth protocol stack for Linux
 *
 *  Copyright (C) 2004-2006  Marcel Holtmann <marcel@holtmann.org>
 *
 *
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program; if not, write to the Free Software
 *  Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
 *
 */

#ifdef HAVE_CONFIG_H
#include <config.h>
#endif

#include <stdio.h>
#include <errno.h>
#include <ctype.h>
#include <unistd.h>
#include <stdlib.h>
#include <malloc.h>
#include <getopt.h>
#include <signal.h>
#include <sys/ioctl.h>
#include <sys/socket.h>

#include <bluetooth/bluetooth.h>
#include <bluetooth/hci.h>
#include <bluetooth/hci_lib.h>

#include <netinet/in.h>

#include "parser/parser.h"

static volatile sig_atomic_t __io_canceled = 0;

static void sig_hup(int sig)
{
}

static void sig_term(int sig)
{
	__io_canceled = 1;
}

static int read_revision(int dd, char *revision, int size)
{
	struct hci_request rq;
	unsigned char req[] = { 0x07 };
	unsigned char buf[46];

	memset(&rq, 0, sizeof(rq));
	rq.ogf    = OGF_VENDOR_CMD;
	rq.ocf    = 0x000e;
	rq.cparam = req;
	rq.clen   = sizeof(req);
	rq.rparam = &buf;
	rq.rlen   = sizeof(buf);

	if (hci_send_req(dd, &rq, 1000) < 0)
		return -1;

	if (buf[0] > 0) {
		errno = EIO;
		return -1;
	}

	if (revision)
		strncpy(revision, (char *) (buf + 1), size);

	return 0;
}

static int enable_sniffer(int dd, uint8_t enable)
{
	struct hci_request rq;
	unsigned char req[] = { 0x00, enable };
	unsigned char buf[1];

	memset(&rq, 0, sizeof(rq));
	rq.ogf    = OGF_VENDOR_CMD;
	rq.ocf    = 0x000e;
	rq.cparam = req;
	rq.clen   = sizeof(req);
	rq.rparam = &buf;
	rq.rlen   = sizeof(buf);

	if (hci_send_req(dd, &rq, 1000) < 0)
		return -1;

	if (buf[0] > 0) {
		errno = EIO;
		return -1;
	}

	return 0;
}

static int enable_sync(int dd, uint8_t enable, bdaddr_t *bdaddr)
{
	struct hci_request rq;
	unsigned char req[] = { 0x01, enable,
				0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
				0x00, 0xfa, 0x00 };

	memcpy(req + 2, bdaddr, 6);

	memset(&rq, 0, sizeof(rq));
	rq.ogf    = OGF_VENDOR_CMD;
	rq.ocf    = 0x000e;
	rq.cparam = req;
	rq.clen   = sizeof(req);

	hci_send_req(dd, &rq, 1000);

	return 0;
}

static char *type2str(uint8_t type)
{
	switch (type) {
	case 0x00:
		return "NULL";
	case 0x01:
		return "POLL";
	case 0x02:
		return "FHS";
	case 0x03:
		return "DM1";
	case 0x04:
		return "DH1";
	case 0x05:
		return "HV1";
	case 0x06:
		return "HV2";
	case 0x07:
		return "HV3";
	case 0x08:
		return "DV";
	case 0x09:
		return "AUX1";
	case 0x0a:
		return "DM3";
	case 0x0b:
		return "DH3";
	case 0x0c:
		return "EV4";
	case 0x0d:
		return "EV5";
	case 0x0e:
		return "DM5";
	case 0x0f:
		return "DH5";
	case 0xff:
		return "ID";
	default:
		return "UNK";
	}
}

static void decode(unsigned char *buf, int count)
{
	struct frame frm;
	uint8_t id, status, channel;
	uint16_t num, len;
	uint32_t time;
	uint8_t type, addr, temp, hdr;
	uint8_t flow, arqn, seqn, hec, llid, pflow;
	uint16_t plen;

	if (count < 7)
		return;

	id = buf[0];
	num = ntohs(bt_get_unaligned((uint16_t *) (buf + 1)));
	len = btohs(bt_get_unaligned((uint16_t *) (buf + 3)));

	status  = buf[5];
	time    = ntohl(bt_get_unaligned((uint32_t *) (buf + 6)));
	channel = buf[10];

	if (len < 8)
		return;

	type = (len < 7) ? 0xff : bt_get_unaligned((uint8_t *) (buf + 11));

	if (type < 2)
		return;

	p_indent(-1, NULL);

	memset(&frm, 0, sizeof(frm));
	frm.data     = buf + 12;
	frm.data_len = count - 12;
	frm.ptr      = frm.data;
	frm.len      = frm.data_len;
	frm.in       = 0;
	frm.master   = 0;
	frm.handle   = 0;
	frm.flags    = 0;

	p_indent(0, &frm);

	printf("BPA: id %d num %d status 0x%02x time %d channel %2d len %d\n",
		id, num, status, time, channel, len - 6);

	if (type < 3) {
		printf("  %s\n", type2str(type));
		raw_dump(1, &frm);
		return;
	}

	addr = bt_get_unaligned((uint8_t *) (buf + 12));
	temp = bt_get_unaligned((uint8_t *) (buf + 13));
	flow = (temp & 0x04) >> 2;
	arqn = (temp & 0x02) >> 1;
	seqn = (temp & 0x01);
	hec  = bt_get_unaligned((uint8_t *) (buf + 14));

	hdr = bt_get_unaligned((uint8_t *) (buf + 20));
	plen  = ((hdr & 0x10) >> 4) | ((hdr & 0x08) >> 2) | (hdr & 0x04) | ((hdr & 0x02) << 2) | ((hdr & 0x01) << 4);
	pflow = ((hdr & 0x20) >> 5);
	llid = ((hdr & 0x80) >> 7) | ((hdr & 0x40) >> 5);
	hdr = bt_get_unaligned((uint8_t *) (buf + 21));
	plen = plen | ((hdr & 0x80) >> 2) | (hdr & 0x40) | ((hdr & 0x20) << 2) | ((hdr & 0x08) << 4);

	p_indent(0, &frm);

	printf("%s: addr 0x%02x flow %d arqn %d seqn %d hec 0x%02x llid %d pflow %d plen %d\n",
		type2str(type), addr, flow, arqn, seqn, hec, llid, pflow, plen);

	if (type == 0x03 && llid == 3) {
		memset(&frm, 0, sizeof(frm));
		frm.data     = buf + 22;
		frm.data_len = plen;
		frm.ptr      = frm.data;
		frm.len      = frm.data_len;
		frm.in       = 0;
		frm.master   = 1;
		frm.handle   = 0;
		frm.flags    = llid;

		lmp_dump(1, &frm);
		return;
	}

	raw_dump(1, &frm);
}

static void process_frames(int dev)
{
	struct sigaction sa;
	struct hci_filter flt;
	unsigned char *buf;
	int dd, size = 2048;

	buf = malloc(size);
	if (!buf) {
		fprintf(stderr, "Can't allocate buffer for hci%d: %s (%d)\n",
						dev, strerror(errno), errno);
		return;
	}

	dd = hci_open_dev(dev);
	if (dd < 0) {
		fprintf(stderr, "Can't open device hci%d: %s (%d)\n",
						dev, strerror(errno), errno);
		free(buf);
		return;
	}

	hci_filter_clear(&flt);
	hci_filter_set_ptype(HCI_VENDOR_PKT, &flt);
	hci_filter_set_ptype(HCI_EVENT_PKT, &flt);
	hci_filter_set_event(EVT_VENDOR, &flt);

	if (setsockopt(dd, SOL_HCI, HCI_FILTER, &flt, sizeof(flt)) < 0) {
		fprintf(stderr, "Can't set filter for hci%d: %s (%d)\n",
						dev, strerror(errno), errno);
		hci_close_dev(dd);
		free(buf);
		return;
	}

	memset(&sa, 0, sizeof(sa));
	sa.sa_flags   = SA_NOCLDSTOP;
	sa.sa_handler = SIG_IGN;
	sigaction(SIGCHLD, &sa, NULL);
	sigaction(SIGPIPE, &sa, NULL);

	sa.sa_handler = sig_term;
	sigaction(SIGTERM, &sa, NULL);
	sigaction(SIGINT,  &sa, NULL);

	sa.sa_handler = sig_hup;
	sigaction(SIGHUP, &sa, NULL);

	while (!__io_canceled) {
		int len;

		len = read(dd, buf, size);
		if (len < 0)
			break;
		if (len < 2)
			continue;

		if (buf[0] == 0x04 && buf[1] == 0xff) {
			if (buf[3] == 0x02) {
				switch (buf[4]) {
				case 0x00:
					printf("Waiting for synchronization...\n");
					break;
				case 0x08:
					printf("Synchronization lost\n");
					__io_canceled = 1;
					break;
				default:
					printf("Unknown event 0x%02x\n", buf[4]);
					break;
				}
			}
		}

		if (buf[0] != 0xff)
			continue;

		decode(buf + 1, len - 1);
	}

	hci_close_dev(dd);

	free(buf);
}

static void usage(void)
{
	printf("bpasniff - Utility for the BPA 100/105 sniffers\n\n");
	printf("Usage:\n"
		"\tbpasniff [-i <dev>] <master-bdaddr>\n");
}

static struct option main_options[] = {
	{ "help",	0, 0, 'h' },
	{ "device",	1, 0, 'i' },
	{ 0, 0, 0, 0}
};

int main(int argc, char *argv[])
{
	struct hci_dev_info di;
	struct hci_version ver;
	char rev[46];
	bdaddr_t bdaddr;
	int dd, opt, dev = 0;

	bacpy(&bdaddr, BDADDR_ANY);

	while ((opt=getopt_long(argc, argv, "+i:h", main_options, NULL)) != -1) {
		switch (opt) {
		case 'i':
			dev = hci_devid(optarg);
			if (dev < 0) {
				perror("Invalid device");
				exit(1);
			}
			break;

		case 'h':
		default:
			usage();
			exit(0);
		}
	}

	argc -= optind;
	argv += optind;
	optind = 0;

	argc -= optind;
	argv += optind;
	optind = 0;

	if (argc < 1) {
		usage();
		exit(1);
	}

	str2ba(argv[0], &bdaddr);

	dd = hci_open_dev(dev);
	if (dd < 0) {
		fprintf(stderr, "Can't open device hci%d: %s (%d)\n",
						dev, strerror(errno), errno);
		exit(1);
	}

	if (hci_devinfo(dev, &di) < 0) {
		fprintf(stderr, "Can't get device info for hci%d: %s (%d)\n",
						dev, strerror(errno), errno);
		hci_close_dev(dd);
		exit(1);
	}

	if (hci_read_local_version(dd, &ver, 1000) < 0) {
		fprintf(stderr, "Can't read version info for hci%d: %s (%d)\n",
						dev, strerror(errno), errno);
		hci_close_dev(dd);
		exit(1);
	}

	if (ver.manufacturer != 12) {
		fprintf(stderr, "Can't find sniffer at hci%d: %s (%d)\n",
						dev, strerror(ENOSYS), ENOSYS);
		hci_close_dev(dd);
		exit(1);
	}

	if (read_revision(dd, rev, sizeof(rev)) < 0) {
		fprintf(stderr, "Can't read revision info for hci%d: %s (%d)\n",
						dev, strerror(errno), errno);
		hci_close_dev(dd);
		exit(1);
	}

	printf("%s\n", rev);

	if (enable_sniffer(dd, 0x01) < 0) {
		fprintf(stderr, "Can't enable sniffer for hci%d: %s (%d)\n",
						dev, strerror(errno), errno);
		hci_close_dev(dd);
		exit(1);
	}

	if (enable_sync(dd, 0x01, &bdaddr) < 0) {
		fprintf(stderr, "Can't enable sync for hci%d: %s (%d)\n",
						dev, strerror(errno), errno);
		enable_sniffer(dd, 0x00);
		hci_close_dev(dd);
		exit(1);
	}

	init_parser(DUMP_EXT | DUMP_VERBOSE, ~0L, 0, DEFAULT_COMPID, -1, -1);

	process_frames(dev);

	if (enable_sync(dd, 0x00, &bdaddr) < 0) {
		fprintf(stderr, "Can't disable sync for hci%d: %s (%d)\n",
						dev, strerror(errno), errno);
		enable_sniffer(dd, 0x00);
		hci_close_dev(dd);
		exit(1);
	}

	if (enable_sniffer(dd, 0x00) < 0) {
		fprintf(stderr, "Can't disable sniffer for hci%d: %s (%d)\n",
						dev, strerror(errno), errno);
		hci_close_dev(dd);
		exit(1);
	}

	hci_close_dev(dd);

	return 0;
}
